scio-core/src/main/java/com/spotify/scio/transforms/DoFnWithResource.java

```diff
@@ -117,14 +119,43 @@ public void setup() {
         resourceId = instanceId + "-" + this.toString();
         break;
     }
-    resources.computeIfAbsent(
+    resources.compute(
         resourceId,
-        key -> {
-          LOG.debug("Creating resource {}", resourceId);
-          return createResource();
+        (key, value) -> {
+          if (value == null) {
+            LOG.debug("Creating resource {}", resourceId);
+            AtomicInteger resourceUsersCounter = new AtomicInteger(1);
+            return Pair.of(resourceUsersCounter, createResource());
+          }
+          else {
+            int newCurrentUsers = value.getLeft().incrementAndGet();
```

Comment:
This happens per element right? So this `AtomicInteger` could be up to total number of elements the `DoFn` has processed, and could overflow? OTOH decrement only happens in `teardown`, so by default once per CPU core. That's not enough to get the counter back to 0.

It seems that for this to work, we need to add a `releaseResource()` method to decrement the counter, and rely on the user to match each `getResource` with a call `releaseResource`. That doesn't seem reliable.

OTOH we can increment the counter in `setup()`? That's usually called once per CPU or `targetParallelism`. Same for `teardown()`. So the counter will be back to 0 when all instances of the `DoFn` are torn down.

Reply:
This happens at the setup of the DoFn, not per element processed. The algorithm is actually as you are describing, increasing at the setup and decrementing at the teardown.

Reply:
Oh duh right! For some reason I remembered the resource was computed lazily in `getResource()`.